src/onnx/parse_qlinearadd.cpp

```diff
@@ -0,0 +1,202 @@
+  A_scale : tensor(float)
+  Input A's scale. It's a scalar, which means a per-tensor/layer quantization.
+
+  A_zero_point (optional) : T
+  Input A zero point. Default value is 0 if it's not specified. It's a scalar, which means a
+  per-tensor/layer quantization.
+
+  B : T
+  Second operand.
+
+  B_scale : tensor(float)
+  Input B's scale. It's a scalar, which means a per-tensor/layer quantization.
+
+  B_zero_point (optional) : T
+  Input B zero point. Default value is 0 if it's not specified. It's a scalar, which means a
+  per-tensor/layer quantization.
+
+  C_scale : tensor(float)
+  Output scale. It's a scalar, which means a per-tensor/layer quantization.
+
+  C_zero_point (optional) : T
+
+  Output zero point. Default value is 0 if it's not specified. It's a scalar, which means a
+  per-tensor/layer quantization.
+
+  Outputs
+  C : T
+  Result, has same element type as two inputs
+
+  Type Constraints
+  T : tensor(uint8), tensor(int8)
+  Constrain input and output types to 8 bit signed and unsigned tensors.
+
+*/
+
+struct parse_qlinearadd : op_parser<parse_qlinearadd>
+{
+    std::vector<op_desc> operators() const { return {{"QLinearAdd"}}; }
+
+    // basic type checking for QLinearAdd Operator
+    void check_inputs(const std::vector<instruction_ref>& args) const
+    {
+        if(args.size() < 7)
+            MIGRAPHX_THROW("QLINEARADD: missing inputs");
+
+        const auto& in_a = args[0];
+        const auto& in_b = args[3];
+
+        auto sh_a = in_a->get_shape();
+        auto sh_b = in_b->get_shape();
+        if(sh_a != sh_b)
+            MIGRAPHX_THROW("QLINEARADD: mismatched input shapes");
+
+        auto type_a = sh_a.type();
+        auto type_b = sh_b.type();
+        if(type_a != migraphx::shape::int8_type and type_a != migraphx::shape::uint8_type)
+            MIGRAPHX_THROW("QLINEARADD: unsupported input type");
+        if(type_b != migraphx::shape::int8_type and type_b != migraphx::shape::uint8_type)
+            MIGRAPHX_THROW("QLINEARADD: unsupported input type");
+    }
+
+    instruction_ref bcast_scalar_instr(const migraphx::shape& shape_out,
+                                       const instruction_ref& arg_in,
+                                       const onnx_parser::node_info& info) const
+    {
+        auto bcast_instr_out = info.add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", shape_out.lens()}}), arg_in);
+        return bcast_instr_out;
+    }
+
+    // This method is to prep for quantizelinear or dequantizelinear operation for
+    // either the broadcasting of weight-scale or zero-points of qlinearadd operator
+    // outputs: operator op (inputs x, broadcasted: scale (float) & zero_pt (8-bit))
+    instruction_ref bcast_qdq_instr(const std::string& op_name,
+                                    const instruction_ref& x_in,
+                                    const instruction_ref& arg_fscale,
+                                    const instruction_ref& arg_z_pt,
```

Comment:
Pass `instruction_ref` by value.

Reply:
We could pull the common functionality (amongst quant operators, e.g. Qlinearconv etc.) bcast_qdq_instr  into a separate file.. Rather than have 4 copies of it..

Reply:
Yes thats a good idea. You can add it under `include/migraphx/onnx/`.

Reply:
It also would be better to rename the function to `insert_linear_instructions`.

Reply:
Another idea, would be to use the same parser class for all these operators. The parser class can list multiple operators:

```cpp
std::vector<op_desc> operators() const 
{ 
    return {{"QLinearAdd", "add"},
            {"QLinearConv", "convolution"}}; 
}
```

The `op_desc.onnx_name` will be the first name(ie `QLinearAdd`) and `op_desc.op_name` will be the second (ie `add`). You can see parse_reduce_op.cpp as an example:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/onnx/parse_reduce_op.cpp#L93

I dont know if that will be easier since it looks this might share more than just one function.

Reply:
I have looked at parse_pooling.cpp as an example. Not too much commonality amongst these quant operators except the need to quantize/dequantize. They are all of different classes, with qdq as the commonality..

Reply:
Moved to another file, in the subdir you suggested. Thanks.